src/useQueue/useQueue.ts

```diff
@@ -0,0 +1,52 @@
+import { useState } from 'react';
+
+export interface QueueMethods<T> {
+  /**
+   * Function for adding an item to to the queue
+   * @param item The item to be added
+   */
+  add: (item: T) => void;
+  /**
+   * Function to remove and return an item from the queue
+   */
+  remove: () => T;
+  /**
+   * The first item in current queue
+   */
+  first: T;
+  /**
+   * The last item in current queue
+   */
+  last: T;
+  /**
+   * The current size of te queue
+   */
+  size: number;
+}
+
+/**
+ * A react state hook implements a simple first-in first-out (FIFO) queue.
+ * @param initialValue an array for the initial value of the queue
+ */
+export const useQueue = <T>(initialValue: T[] = []): QueueMethods<T> => {
+  const [state, set] = useState(initialValue);
+  return {
+    add: (value) => {
+      set((queue) => [...queue, value]);
+    },
+    remove: () => {
+      const result = state[0];
+      set((current) => current.slice(1, current.length));
+      return result;
+    },
+    get first() {
+      return state[0];
+    },
+    get last() {
+      return state[state.length - 1];
+    },
+    get size() {
+      return state.length;
+    },
+  };
+};
```

Comment:
The returned functions here are not referentially stable. Their reference will change between renders, which will cause problems, for example, an infinite loop if used inside of `useEffect`.

You can use `useMemo` to achieve referential stability. You can also take a look at some of our other hooks to see how they've achieved this. For example, see [useAsync](https://github.com/react-hookz/web/blob/master/src/useAsync/useAsync.ts).

I pushed a test to this PR, which tests for referential equality of the returned functions.

Reply:
Good call out! I didn't even think of this. Sorry I didn't check the "allow edits from maintainers" box. It's checked now. I went ahead and added this test as well as a Memoization of the return.